cpg_workflows/stages/realign_genotype_with_dragen.py

```diff
@@ -0,0 +1,357 @@
+)
+from cpg_workflows.targets import SequencingGroup
+from cpg_workflows.workflow import SequencingGroupStage, StageInput, StageOutput, stage
+
+if TYPE_CHECKING:
+    from hailtop.batch.job import BashJob, PythonJob
+
+DRAGEN_VERSION: Final = config_retrieve(['ica', 'pipelines', 'dragen_version'])
+GCP_FOLDER_FOR_ICA_PREP: Final = f'ica/{DRAGEN_VERSION}/prepare'
+GCP_FOLDER_FOR_RUNNING_PIPELINE: Final = f'ica/{DRAGEN_VERSION}/pipelines'
+GCP_FOLDER_FOR_ICA_DOWNLOAD: Final = f'ica/{DRAGEN_VERSION}/output'
+ICA_REST_ENDPOINT: Final = 'https://ica.illumina.com/ica/rest'
+
+
+coloredlogs.install(level=logging.INFO)
+
+
+def calculate_needed_storage(
+    cram: str,
+    bucket_name: str,
+) -> str:
+    logging.info(f'Checking blob size for {cram}')
+    storage_client = storage.Client()
+    gcp_bucket = storage_client.bucket(bucket_name=bucket_name)
+    blob_to_upload_size_bytes: int = gcp_bucket.get_blob(blob_name=f'{cram}').size
+    storage_size: int = ceil(ceil((blob_to_upload_size_bytes / (1024**3)) + 3) * 1.2)
+    logging.info(f'Calculated storage is {storage_size}Gi.')
+    return f'{storage_size}Gi'
+
+
+# No need to register this stage in Metamist I think, just ICA prep
+@stage(analysis_keys=['cram_fid', 'cram_index_fid', 'analysis_output_fid'])
+class PrepareIcaForDragenAnalysis(SequencingGroupStage):
+    """Set up ICA for a single realignment run.
+
+    Creates a file ID for both the CRAM and CRAI file to upload to.
+    Creates a folder ID for the Dragen output to be written into.
+
+    """
+
+    def expected_outputs(self, sequencing_group: SequencingGroup) -> dict[str, cpg_utils.Path]:
+        sg_bucket: str = f'{get_path_components_from_gcp_path(str(sequencing_group.cram))["bucket"]}'
+        output_dict = {
+            'cram_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}.cram_ica_file_id',
+            ),
+            'crai_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}.cram.crai_ica_file_id',
+            ),
+            'analysis_output_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}_dragen_output_folder_id',
+            ),
+        }
+        return output_dict
+
+    def queue_jobs(self, sequencing_group: SequencingGroup, inputs: StageInput) -> StageOutput | None:
+        cram_path_components: dict[str, str] = get_path_components_from_gcp_path(path=str(object=sequencing_group.cram))
+        cram: str = cram_path_components['file']
+        bucket_name: str = cram_path_components['bucket']
+        logging.info(bucket_name)
+
+        prepare_ica_job: PythonJob = get_batch().new_python_job(
+            name='PrepareIcaForDragenAnalysis',
+            attributes=(self.get_job_attrs() or {}) | {'tool': 'ICA'},
+        )
+        prepare_ica_job.image(image=image_path('cpg_workflows'))
+
+        prepare_ica_job.call(
```

Comment:
As a general note having an explicit connection between the output of `expected_outputs` (generating the result paths) and the process that writes those files makes a lot of sense to me. Here you're building the paths to output files in `expected_outputs` from components, then building the same path elsewhere from components in a different way. 

The end result is that the outputs are created, but it's difficult to trace why/how they're created at the right paths due to the indirection.

Reply:
This is probably the biggest deviation from every other stage in this pipeline framework. I'm guessing that's because the helper methods you've created (like `register_output_to_gcp`) require various parts of filename to use the google storage client, when you could use the Path object on the exact Path objects which are being created in `expected_ouputs`, e.g.

```
output_path = self.expected_outputs()

# do whatever to get the result
workflow_id = <some method call>

# open the Path for writing directly
with output_path.open('w') as handle:
    handle.write(workflow_id)
```